### PR TITLE
ECS: Add Storage Objects

### PR DESCRIPTION
**Pull request by christopher-buss**
##### Description

Adds storage objects. For now the only possible object is `useEvent`, but this gives us the ability to add further changes in the future.

### CODE REVIEW

File: src/lib/ecs/storage/event.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	 * ```
+	 */
+	function* items(): Generator<InferSignalParameters<E>> {
+		while (queue.size() > 0) {
+			const args = queue.pop();
+			if (args) {
+				yield args as InferSignalParameters<E>;
+			}
+		}
+	}
```
Comment (RigidStudios) [inline src/lib/ecs/storage/event.ts:84]: Standardize this to `.iter()` or give us a separate `.signals()` or something of the sort.

File: src/lib/ecs/system.ts [hunk 2]
```diff
@@ -363,7 +391,9 @@ export class SystemManager {
 				system.recentErrors = new Map();
 			}
 
-			const recentError = `System: ${system.name} errored! ${result} + \n ${debug.traceback}`;
+			const recentError = `System: ${tostring(
```
Comment (RigidStudios) [inline src/lib/ecs/system.ts:394]:
```suggestion
			const recentError = `System ${tostring(
```

File: src/lib/ecs/system.ts [hunk 3]
```diff
@@ -475,7 +503,7 @@ export class SystemManager {
 			system.dt = os.clock() - system.lastCalled;
 			system.lastCalled = os.clock();
 
-			debug.profilebegin("system: " + system.name);
+			debug.profilebegin("system: " + tostring(getmetatable(system)));
```
Comment (RigidStudios) [inline src/lib/ecs/system.ts:506]: I don't like how our performance enforcer contains a concatenation 🤔
As long as people name their systems well, simply using the system name should be sufficient.

File: src/lib/util/connection-util.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+					: S["connect"] extends Callback
+					? S["connect"]
+					: S["on"] extends Callback
+					? S["on"]
+					: S["do"] extends Callback
+					? S["do"]
+					: never
+			>[0]
+	  >
+	: never;
```
Comment (RigidStudios) [inline src/lib/util/connection-util.ts:18]: Remove the lowercase connect ones, haven't really seen that stuff in 5-ish years so I'd rather not support bad code

File: src/lib/util/connection-util.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+export interface SignalLike {
+	Connect?(callback: Callback): ConnectionLike;
+	connect?(callback: Callback): ConnectionLike;
+	do?(callback: Callback): ConnectionLike;
+	on?(callback: Callback): ConnectionLike;
+}
+
+const EVENT_CONNECT_METHODS = ["Connect", "do", "on", "connect"];
+const EVENT_DISCONNECT_METHODS = ["Disconnect", "Destroy", "disconnect", "destroy"];
```
Comment (RigidStudios) [inline src/lib/util/connection-util.ts:35]:
```suggestion
const EVENT_CONNECT_METHODS: Array<keyof SignalLike> = ["Connect", "do", "on", "connect"];
const EVENT_DISCONNECT_METHODS: Array<keyof ConnectionLike> = ["Disconnect", "Destroy", "disconnect", "destroy"];
```

File: src/lib/util/connection-util.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	 */
+	export function connect(event: SignalLike, callback: Callback): ConnectionLike {
+		if (typeIs(event, "RBXScriptSignal")) {
+			return event.Connect(callback);
+		}
+
+		for (const method of EVENT_CONNECT_METHODS) {
+			if (type(event[method as keyof SignalLike["Connect"]]) === "function") {
+				return event[method as keyof SignalLike]!(callback);
+			}
```
Comment (RigidStudios) [inline src/lib/util/connection-util.ts:58]:
```suggestion
			if (type(event[method]) === "function") {
				return event[method]!(callback);
			}
```

File: src/lib/util/connection-util.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	 */
+	export function disconnect(connection: ConnectionLike): void {
+		if (typeIs(connection, "RBXScriptConnection")) {
+			return connection.Disconnect();
+		}
+
+		for (const method of EVENT_DISCONNECT_METHODS) {
+			if (type(connection[method as keyof ConnectionLike["Destroy"]]) === "function") {
+				connection[method as keyof ConnectionLike]!();
+			}
```
Comment (RigidStudios) [inline src/lib/util/connection-util.ts:78]:
```suggestion
			if (type(connection[method]) === "function") {
				connection[method]!();
			}
```

**Review by RigidStudios — CHANGES_REQUESTED** [review]

**Review by christopher-buss — COMMENTED** [review]
modifications based on suggestions

**Pull request merged**